src/main/java/duke/Ui.java

```diff
@@ -0,0 +1,136 @@
+                + "  > ^ <    What can I do for you? :)\n" + PARTITION + "\n";
+        printMessage(startMessage);
+    }
+
+    String readCommand() throws DukeException {
+        Scanner sc = new Scanner(System.in);
+        String input = sc.nextLine();
+        validate(input.toLowerCase());
+        return input;
+    }
+
+    public void showLine() {
+        System.out.println(PARTITION);
+    }
+
+    public void printMessage(String message) {
+        System.out.println(message);
+    }
+
+    /**
+     * Converts list containing tasks to string for printing.
+     * @param list list of user's saved tasks.
+     * @return current saved tasks in string form.
+     */
+    public String listToString(List<Task> list) {
+        if (list.size() == 0) {
+            return "List is empty ~\n";
+        }
+        StringBuilder output = new StringBuilder();
+        output.append("Here are the tasks in your list: \n");
+        for (int i = 0; i < list.size(); i++) {
+            output.append(i + 1).append(". ").append(list.get(i)).append("\n");
+        }
+        return output.toString();
+    }
+
+    /**
+     * Converts list containing tasks with specific text to string for printing.
+     * @param list list of user's saved tasks.
+     * @param text keyword to find in task description.
+     * @return matching tasks in string form.
+     */
+    public String listToStringWithText(List<Task> list, String text) {
+        boolean foundMatchingTask = false;
+        if (list.size() == 0) {
+            return "List is empty ~\n";
+        }
+        StringBuilder output = new StringBuilder();
+        output.append("Here are the matching tasks containing \"" + text + "\":\n");
+        for (int i = 0; i < list.size(); i++) {
+            if (list.get(i).getDescription().contains(text)) {
+                output.append(i + 1).append(". ").append(list.get(i)).append("\n");
+                foundMatchingTask = true;
+            }
+        }
+        if (!foundMatchingTask) {
+            return "No matching tasks with \"" + text + "\" :(";
+        }
+        return output.toString();
+    }
+
+    void showLoadingError() {
+        System.out.println("Error in loading task :( New task list created!");
+    }
+
+    public void showError(String message) {
+        System.out.println(message);
+    }
+
+    void validate(String input) throws DukeException {
+        String[] commandSegments = input.split(" ", 2);
+        String mainCommand = commandSegments[0].toLowerCase().trim();
+
+        String[] allCommands = {"list", "bye", "todo", "deadline", "event", "mark", "unmark", "delete", "find"};
+        if (!Arrays.asList(allCommands).contains(mainCommand)) {
+            //handles invalid commands
+            throw new DukeException(mainCommand + "? (´･_･`) I don't know what that means\n");
+        }
+        String[] commandsWithDescription = {"todo", "deadline", "event", "find"};
```

Comment:
adding a comment header and leaving a blank line would help as well, so that the reader understands that this handles the scenario where certain commands without a body are meaningless and hence should throw an error. 

Perhaps other commands such as mark/unmark without a integer attached to it could be meaningful as well